File: app/src/main/java/org/nervos/neuron/activity/transactionlist/model/TokenDescribeModel.java
package org.nervos.neuron.activity.transactionlist.model;

import android.text.TextUtils;

import com.google.gson.Gson;

import org.nervos.neuron.item.EthErc20TokenInfoItem;
import org.nervos.neuron.util.url.HttpUrls;
import org.nervos.neuron.service.http.HttpService;
import org.nervos.neuron.util.AddressUtil;
import org.web3j.crypto.Keys;

import java.io.IOException;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.Request;
import okhttp3.Response;

/**
 * Created by BaojunCZ on 2018/10/9.
 */
public class TokenDescribeModel {

    private TokenDescribeModelImpl listener;

    public TokenDescribeModel(TokenDescribeModelImpl listener) {
        this.listener = listener;
    }

    public void get(String address) {
        if (AddressUtil.isAddressValid(address))
            address = Keys.toChecksumAddress(address);
        String url = String.format(HttpUrls.TOKEN_DESC, address);
        final Request ethRequest = new Request.Builder().url(url).build();
        Call ethCall = HttpService.getHttpClient().newCall(ethRequest);
        ethCall.enqueue(new Callback() {
            @Override
            public void onFailure(Call call, IOException e) {
                listener.error();
            }

            @Override
            public void onResponse(Call call, Response response) {
                try {
                    String body = response.body().string();
                    response.close();
                    if (!TextUtils.isEmpty(body)) {
                        EthErc20TokenInfoItem item = new Gson().fromJson(body, EthErc20TokenInfoItem.class);
                        listener.success(item);
                    } else {
                        listener.error();
                    }
                } catch (Exception e) {
                    listener.error();
                    e.printStackTrace();
                }
            }
        });
    }

    public interface TokenDescribeModelImpl {
        void success(EthErc20TokenInfoItem item);

        void error();
    }

}

File: app/src/main/java/org/nervos/neuron/view/tokenprofile/TokenDescribeModel.kt
package org.nervos.neuron.view.tokenprofile

import android.text.TextUtils
import com.google.gson.Gson
import okhttp3.Request
import org.nervos.neuron.item.EthErc20TokenInfoItem
import org.nervos.neuron.service.http.HttpService
import org.nervos.neuron.util.AddressUtil
import org.nervos.neuron.util.url.HttpUrls
import org.web3j.crypto.Keys
import rx.Observable
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers

/**
 * Created by BaojunCZ on 2018/10/9.
 */
class TokenDescribeModel {

    operator fun get(address: String): Observable<EthErc20TokenInfoItem?>? {
        return Observable.fromCallable {
            var mAddress = address
            if (AddressUtil.isAddressValid(address))
                mAddress = Keys.toChecksumAddress(address)
            mAddress
        }
                .subscribeOn(Schedulers.newThread())
                .flatMap {
                    val url = String.format(HttpUrls.TOKEN_DESC, it)
                    val request = Request.Builder().url(url).build()
                    val call = HttpService.getHttpClient().newCall(request)
                    val res = call.execute().body()!!.string()
                    if (!TextUtils.isEmpty(res)) {
                        val item = Gson().fromJson(res, EthErc20TokenInfoItem::class.java)
                        Observable.just(item)
                    } else {
                        Observable.just(null)
                    }
                }.observeOn(AndroidSchedulers.mainThread());
    }
}
